# Expressions

## Type of Expressions

**An expression describes a computation and evaluates to a value.**

Expressions are not something particular to computer science. Mathematicians have been describing how to compose different numbers together for a long time, they invented a lot of different ways to describe computation using expressions:
<img src='expression.jpg' width = 600/>

Turns out one of them is a generalization of all the others, and it's all we need: **function call notation**, `f(x)`.
<img src = 'function.jpg' width = 600/>
Instead of using the symbols, subscripts, superscripts, vertical bars, etc., computer scientists write down everything using `function call notation`.

## Call Expressions in Python

**All expressions can use function call notation**

## Demo

Here we type an expression:

In [1]:
2015

2015

And it shows the value. If we write the expression in 2 different values combined together by addition such as below,

In [2]:
2000 + 15

2015

The result value is what we get by combining the value `2000` and `15`.

Python is also a powerful calculator. We can multiply numbers,

In [4]:
2 * 3

6

Or nesting,

In [5]:
2 * ((3 + 4) * 5)

70

Or we can do the following,

In [13]:
1 * 2 * ((3 * 4 * 5 // 6) ** 3) + 7 + 8

2015

`1 * 2 * ((3 * 4 * 5 // 6) ** 3) + 7 + 8` is an expression, while `2015` is its value

What about call expressions? Call expressions have a special form where we name a function that we want to call. In this case, the `max` function takes in expressions `2` and `4` that will give us the value to which `max` will be applied.

In [14]:
max(2, 4)

4

And we can do the same thing for `min`,

In [15]:
min(-2, 50000)

-2

The whole `max(2, 4)` is a call expression.

Earlier, we claimed that everything can be expressed using a call expression. This is including multiplication, division, addition, etc.

In addition to the symbol `+`, there's a function `add`. This function is not readily available, as we have to write the following to obtain access to `add` and `mul`,

In [2]:
from operator import add, mul

Now we can add together `2` and `3` using call expression.

In [17]:
add(2, 3)

5

And we can multiply together `2` and `3` to get 6.

In [18]:
mul(2, 3)

6

Call expressions can combine multiple values together by using lots of commas,

In [19]:
max(1, 2, 3, 4, 5)

5

Call expressions can also be nested within one another. In this case, we multiply the result of adding `2` and the result of multiplying 4 and 6 with the result of adding `3` and `5`.

In [20]:
mul(add(2, mul(4, 6)), add(3, 5))

208

In call expressions, we don't need to memorize the order of operations. The nesting structure of the expression itself tells exactly what get multiplied before it gets added. For example, Python has to multiply `4` and `6` first, then add the result to `2`.

One other important thing to remember is that multiplication and division comes first before addition and subtraction.

# Anatomy of a Call Expression
Now we're going to talk about the anatomy of a call expression. Here is an example: `add(2, 3)`

In [21]:
add(2, 3)

5

The whole `add(2, 3)` is a call expression. This call expression contains other expressions.

1. Everything that comes before the opening parentheses `(` is `Operator Subexpression`. Subexpression means an expression within an expression.
<img src = 'operator.jpg' width = 500/>

2. The `operands` are separated by commas `,` within parentheses `( )`.
<img src = 'operand.jpg' width = 500/>

**Operators and operands are also expressions**. They altogether evaluate to values.

The way call expressions are evaluated is called **evaluation procedure**.

Programming languages interpret expressions by applying evaluation procedures. They apply the same procedures over and over again.

The **evaluation procedure** for call expressions work like the following,
1. Evaluate the operator and then the operand subexpressions
    * We have the `function` `add` from the `operator` and `argument`s `2` and `3` from the `operand`s
<img src = 'evaluation.jpg' width = 500/>
2. **Apply**:
    * The `function` that is the value of the `operator subexpression` to
    * The `arguments` that are the values of the operand subexpression. 
    
All of this is just to add `2` and `3` together. We go through the steps in such detail to show that the same exact evaluation procedure can be used many times to evaluate more complicated, nested expressions.

## Evaluating Nested Expressions

In [4]:
mul(add(4, mul(4, 6)), add(3, 5))

224

When we run the cell above, Python just give out the result `224`. How did Python do that? Python applied the same evaluation procedure as before. 

Python recognizes that the whole code `mul(add(4, mul(4, 6)), add(3, 5))` is a call expression. 

At first, Python evaluates the `operator`, which is the function `mul` that multiplies,
<img src = 'mul.jpg' width = 550/>
Then Python evaluates the first `operand`, `add(4, mul(4, 6))`, which is another call expression! Thus, Python applies the same evaluation procedure to this operand,
<img src = 'operand_2.jpg' width = 300/>


## `add(4, mul(4, 6))`
In this call expression, Python evaluates the `operator`, the `add` function, and the operands:
1. The number `4`
2. The call expression `mul(4, 6)`
<img src = 'add_4_mul_4_6.jpg' width = 400/>

We have a procedure for evaluating call expressions: evaluate the operator and the operand subexpressions. From there, we get:
1. The function `mul` that multiplies 
2. The numbers `4` and `6`. 

Since we have the operands evaluated, we can apply `mul` to `4` and `6` to obtain 24. 
<img src = '24.jpg' width = 500/>
`24` is the value of the call expression `mul(4, 6)`

Now that we have evaluated all the operands from the subexpression `add(4, mul(4, 6))`, 
<img src = 'add_4_24.jpg' width = 500/>
We can apply `add` to `4` and `24`,
<img src = '28.jpg' width = 500/>

## Back to the main call expression
<img src = 'back_to_main.jpg' width = 500/>

Now that we have the following, we still need to evaluate the second operand. By applying the same evaluation procedure as what we have done so far, we can obtaion the value of the call expression `224`

<img src = '224.jpg' width = 500/>

This whole diagram is called an **expression tree**. It is an illustration of everything that happens within the computer as it evaluates this nested call expression.
<img src = 'expression_tree.jpg' width = 600/>

The important thing in this call expression evaluation is the order in which expressions are evaluated. We have to know that the expression `add(4, mul(4, 6))` evaluates to `28` and the expression `add(3, 5)` evaluates to 8 to finish the computation. 
<img src = 'whole.jpg' width = 700/>

Above, `28` is both **value of subexpression `add(4, mul(4, 6))`** and **the first argument to `mul`**